### ISSUE TITLE
Generator-based declarator phase in System format

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#1826]
From https://github.com/babel/proposals/issues/44#issuecomment-393986595, using generators as a baseline avoids the need for manual hoisting, while supporting better TDZ handling.

The example from there adjusted to separate into a two-phase "declare" and "execute" might look like:

```js

System.register(["foo"], function* (_export, _context) {
  "use strict";

  _export(0, function (_foo) {
    var _exportObj = {};

    for (var _key in _foo) {
      if (_key !== "default" && _key !== "__esModule") _exportObj[_key] = _foo[_key];
    }

    _export(_exportObj);
  });
  _export("foo", foo);

  // indicates end of declaration phase
  yield;

  console.log(val);

  let val = 4;
  function foo() {
    console.log(val);
  }

  // indicates end of execution phase, optionally returning execution promise
  yield Promise.resolve().then(function () { ... })
});
```

**Comment by guybedford** [comment]
Main issue here is that top-level await still needs hoisting in this approach... which seems to defeat the purpose.

**Comment by loganfsmyth** [comment]
Could you clarify how it would defeat it? I think I'm missing what you're thinking.

**Comment by guybedford** [comment]
So with the top-level await support I just shared in https://github.com/babel/proposals/issues/44#issuecomment-393978818, for example the following:

```js
import { x } from 'foo';

export function getValue () {
  console.log(x);
  console.log(p);
}

await thing();
let p = 5;
```

would become in the generator form:

```js
System.register(['foo'], function* (_export, _context) {
  "use strict";

  var x;
  _export(0, function (_foo) {
    x = _foo.x;
  });

  function getValue () {
    console.log(x);
    console.log(p);
  }
  _export('getValue', getValue);

  // indicates end of declaration phase
  yield;

  // top-level await handled by returning a promise from execution yield
  yield Promise.resolve()
  .then(async function () {
    await thing();

    // BUG: Should have hoisted "p", using same hoisting logic we have currently
    let p = 5;
  });
});
```

**Comment by loganfsmyth** [comment]
Oh gotcha, I was thinking it would be
```
System.register(["foo"], function* (_export, _context) {
  "use strict";

  _export(0, function (_foo) {
    var _exportObj = {};

    for (var _key in _foo) {
      if (_key !== "default" && _key !== "__esModule") _exportObj[_key] = _foo[_key];
    }

    _export(_exportObj);
  });
  _export("foo", foo);

  // indicates end of declaration phase
  yield;

  console.log(val);

  let val = 4;
  function foo() {
    console.log(val);
  }

  yield thing();
  let p = 5;
});
```
With SystemJS being responsible for suspending execution until the import was resolved.

**Comment by guybedford** [comment]
Right, so doing a sort of mini regenerator-style transform inline? That should be possible, but my worry is handling the more advanced cases with try-catch etc. The other thing with that is that a simple async function output in future isn't possible then as well (when async functions are the new baseline, likely matching the compatibility paths this is looking at in future anyway...).

Feel free to copy me in on any bugs with the module format btw.

**Comment by loganfsmyth** [comment]
> That should be possible, but my worry is handling the more advanced cases with try-catch etc.

Fair enough. I don't think it ends up being particularly complicated once it's implemented, and you could always look at how Babel does it, though.

> The other thing with that is that a simple async function output in future isn't possible then as well

Is that the case? If calling the function synchronously returns an object with `execute` it's the current format, if it returns an object with `.next` it's a generator, and if it returns a thenable, it's an async function?

Also, I guess now that I think about it, you could also go with async functions instead of generators using something like

```
System.register(["foo"], async function (_export, _context) {
  "use strict";

  _export(0, function (_foo) {
    var _exportObj = {};

    for (var _key in _foo) {
      if (_key !== "default" && _key !== "__esModule") _exportObj[_key] = _foo[_key];
    }

    _export(_exportObj);
  });
  _export("foo", foo);

  // indicates end of declaration phase
  await _context.ready();

  console.log(val);

  let val = 4;
  function foo() {
    console.log(val);
  }

  await thing();
  let p = 5;
});
```
Th SystemJS would resolve the `ready()` promise once overall execution was ready to go.

**Comment by guybedford** [comment]
> Fair enough. I don't think it ends up being particularly complicated once it's implemented, and you could always look at how Babel does it, though.

Definitely it can be done, but it is some work and a shame to expect the helper functions to be needed in the transform for all compilers.

> Is that the case? If calling the function synchronously returns an object with execute it's the current format, if it returns an object with .next it's a generator, and if it returns a thenable, it's an async function?

The issue with this is the divergence of the code paths. We don't want to have different code paths for top-level await v normal output as this will add to the transform complexity. We only want to tolerate such divergence if it is towards a format that can handle all scenarios.

> Also, I guess now that I think about it, you could also go with async functions instead of generators using something like

This is a really great suggestion too! I nearly got behind this one too but there's an issue here in that calling the resolver for `_context.ready` still results in an event loop tick before the `console.log(val)` phase, and we need execution to be perfectly synchronous to match exact semantics.

**Comment by loganfsmyth** [comment]
> The issue with this is the divergence of the code paths.

Totally fair.

> still results in an event loop tick

Ah, of course, good call.

**Comment by guybedford** [comment]
@loganfsmyth just realised this works fine with async generators -

```js
System.register(["foo"], async function* (_export, _context) {
  "use strict";

  _export(0, function (_foo) {
    var _exportObj = {};

    for (var _key in _foo) {
      if (_key !== "default" && _key !== "__esModule") _exportObj[_key] = _foo[_key];
    }

    _export(_exportObj);
  });
  _export("foo", foo);

  // indicates end of declaration phase
  yield;

  console.log(val);

  let val = 4;
  function foo() {
    console.log(val);
  }

  await thing();
  let p = 5;
});
```

where the `.next()` on an async generator, while returning a promise will actually call the sync stuff in sync correctly.

I don't know if I have the bandwidth to do the Babel side of this myself for quite a while yet, but if you're interested in working on it I can happily integrate the SystemJS side. Let me know.

**Comment by guybedford** [comment]
The only hard bit left over is determining whether the main execution body is sync or async... we may need some kind of export flag to indicate that - 

```js
_export(-1, true);
```

**Comment by guybedford** [comment]
@Jasernator note this is ongoing tracking issue on a new modern System format v2 which may interest you.

**Comment by guybedford** [comment]
@Jamesernator note this is ongoing tracking issue on a new modern System format v2 which may interest you.
